### PR TITLE
make the ping queue use a more compact data structure

### PR DESCRIPTION
**Pull request by arvidn**
and reduce some IPv4/IPv6 code duplication.

### CODE REVIEW

File: main.cpp [hunk 1]
```diff
@@ -390,19 +390,6 @@ struct bound_socket
 	char packet[1500];
 };
 
-// this is the type of each node queued up
-// to be pinged at some point in the future
-struct queued_node_t
```
Comment (arvidn) [inline main.cpp]: I moved this down a bit and stripped out fields that aren't used externally

File: main.cpp [hunk 2]
```diff
@@ -442,97 +429,110 @@ struct ip_set
...
 };
 
+struct queued_node_t
+{
+	udp::endpoint ep;
+	int sock_idx;
+};
+
+template <typename Address>
 struct ping_queue_t
```
Comment (arvidn) [inline main.cpp]: while working on this increasingly more fields started having an IPv4 and IPv6 version, so making it a template made that a lot cleaner. I also moved in the queue entry type as an inner class

File: main.cpp [hunk 3]
```diff
@@ -442,97 +429,110 @@ struct ip_set
...
+	struct queue_entry_t
+	{
+		typename Address::bytes_type addr;
+
+		// the index of the socket this node was seen on, and should be pinged
+		// back via
+		std::uint32_t sock_idx:2;
+
+		// expiration in seconds (relative to the ping_queue creation time)
+		std::uint32_t expire:30;
```
Comment (arvidn) [inline main.cpp]: I suppose technically there will always be  bytes padding here, so the bitfield may be a bit redundant

Comment (ssiloti) [inline main.cpp]: Yeah, as-is the bitfield is just generating slower code for no gain.

**Comment by ssiloti** [comment]
lgtm

**Pull request merged**